benchmarks/cpp/src/qsort.cpp

```diff
@@ -0,0 +1,73 @@
+#include <algorithm>
+#include <asio.hpp>
+#include <vector>
+#include <iostream>
+#include <random>
+
+using namespace std::chrono;
+using asio::awaitable;
+using asio::co_spawn;
+using asio::detached;
+
+constexpr int partition(std::vector<int>& arr, unsigned long start, unsigned long end) noexcept {
+    const auto pivot = static_cast<unsigned long>(end - 1);
+    int i = start;
+    for (unsigned long j = start; j < pivot; ++j) {
+        if (arr[j] <= arr[pivot]) {
+            std::iter_swap(arr.begin() + i, arr.begin() + j);
+            ++i;
+        }
+    }
+    std::iter_swap(arr.begin() + i, arr.begin() + pivot);
+    return i;
+}
+
+constexpr void insertionSort(std::vector<int>& arr, unsigned long start, unsigned long end) noexcept {
+    for (unsigned long i = start + 1; i < end; ++i) {
+        for (unsigned long n = i; n > start && arr[n] < arr[n - 1]; --n) {
+            std::iter_swap(arr.begin() + n, arr.begin() + n - 1);
+        }
+    }
+}
+
+void shuffle(std::vector<int>& arr) {
+    std::mt19937 rng(std::random_device{}());
+    std::shuffle(arr.begin(), arr.end(), rng);
+}
+
+awaitable<void> quickSort(asio::io_context& ctx, std::vector<int>& arr, unsigned long start, unsigned long end) noexcept {
+    if (end - start <= 32) {
+        insertionSort(arr, start, end);
+        co_return;
+    }
+
+    const auto pivot = partition(arr, start, end);
+    co_await quickSort(ctx, arr, start, start + pivot);
+    co_await quickSort(ctx, arr, start + pivot, end);
+}
+
+int main() {
+    std::vector<int> arr(10'000);
+
+    std::cout << "filling" << std::endl;
+    std::iota(arr.begin(), arr.end(), 0);
+
+    std::cout << "shuffling" << std::endl;
+    shuffle(arr);
+
+    std::cout << "running" << std::endl;
+    const auto start = high_resolution_clock::now();
+    asio::io_context ctx;
+    co_spawn(ctx, [&]() -> awaitable<void> {
+        co_await quickSort(ctx, arr, 0, arr.size());
+    }, detached);
+    ctx.run();
```

Comment:
Neat. Is there a way to make the io_context now use multiple threads since scheduling is fixed? Not sure if it does so by default.

Reply:
[`asio::thread_pool`](https://think-async.com/Asio/asio-1.24.0/doc/asio/reference/thread_pool.html)

Reply:
I believe that using a thread pool to execute the recursive calls to quickSort instead of fibers may result in slower execution because of the overhead of creating and managing threads.
But, I am still struggling quite a bit here about.... Studying about `asio` is not easy at all (esoteric lib)!

Reply:
I'ts possible add threads on context: `asio::io_context ctx{N};` to N threads!

Reply:
Also, I have tried to use thread pooling, but an error occurs when the array is not sorted.
```c++
// main function, after declaring ctx;
asio::thread_pool pool{4}; // specify the number of threads to use
    asio::io_context::work work{ctx}; // keep the io_context running until stopped
    co_spawn(pool, [&]() -> awaitable<void> {
        co_await quickSort(ctx, arr, 0, arr.size());
    }, detached);
    ctx.run();
    pool.join();
``

Reply:
The array not being sorted means that it's not waiting for all spawned quickSorts correctly. Would have to see how quickSort() does the co_spawn (it should use pool, and youd probably need a way to synchronize shutdown of it).

>  using a thread pool to execute the recursive calls to quickSort instead of fibers may result in slower execution because of the overhead of creating and managing threads.

Thread-pools are meant to spawn threads and keep them alive, supposedly amortizing the cost of creating/managing them. These benchmarks are specifically made to put strain on their runtime's thread-pools as that overhead is what is being measured/compared. (There are definitely faster ways to write a parallel quicksort than those in this repo; The rust `rayon::join` version employs those optimizations and I had to use one of its APIs that doesn't to keep what it's comparing fair)

Reply:
Indeed.